data/base.go

```diff
@@ -0,0 +1,256 @@
+package data
+
+import (
+	"fmt"
+	"reflect"
+	"time"
+
+	valid "github.com/tidepool-org/platform/Godeps/_workspace/src/gopkg.in/bluesuncorp/validator.v8"
+	"github.com/tidepool-org/platform/Godeps/_workspace/src/labix.org/v2/mgo/bson"
+
+	"github.com/tidepool-org/platform/validate"
+)
+
+//used for all data types
+var validator = validate.NewPlatformValidator()
+
+func init() {
+	validator.RegisterValidation(timeStringTag, TimeStringValidator)
+	validator.RegisterValidation(timeObjectTag, TimeObjectValidator)
+	validator.RegisterValidation(timeZoneOffsetTag, TimezoneOffsetValidator)
+	validator.RegisterValidation(payloadTag, PayloadValidator)
+	validator.RegisterValidation(annotationsTag, AnnotationsValidator)
+}
+
+type Base struct {
+	//required data
+	_ID      bson.ObjectId `bson:"_id" valid:"mongo,required"`
+	ID       string        `json:"id" bson:"id" valid:"required"`
+	UserID   *string       `json:"userId" bson:"userId" valid:"required"`
+	DeviceID *string       `json:"deviceId" bson:"deviceId" valid:"required"`
+	Time     *string       `json:"time" bson:"time" valid:"timestr"`
+	Type     *string       `json:"type" bson:"type" valid:"required"`
+	UploadID *string       `json:"uploadId" bson:"uploadId" valid:"-"`
+
+	//optional data
+	DeviceTime       *string       `json:"deviceTime,omitempty" bson:"deviceTime,omitempty" valid:"omitempty,timestr"`
+	TimezoneOffset   *int          `json:"timezoneOffset,omitempty" bson:"timezoneOffset,omitempty" valid:"omitempty,timezoneoffset"`
+	ConversionOffset *int          `json:"conversionOffset,omitempty" bson:"conversionOffset,omitempty" valid:"omitempty,required"`
+	ClockDriftOffset *int          `json:"clockDriftOffset,omitempty" bson:"clockDriftOffset,omitempty" valid:"omitempty,required"`
+	Payload          interface{}   `json:"payload,omitempty" bson:"payload,omitempty" valid:"omitempty,payload"`
```

Comment:
Might want pointer here and for Annotations since a nil is different than an empty map or empty array.